File: projectfiles/flameatlab/functions/Syrk_ln_blk_var5.m
% Copyright 2025 The University of Texas at Austin
%
% For licensing information see
%                http://www.cs.utexas.edu/users/flame/license.html
%
% C ← C + A Aᵀ   (lower-stored SYRK, blocked variant)

function [ C_out ] = Syrk_ln_blk_var5( A, AT, C, nb_alg )

  % Partition A by columns, AT by *rows*
  [ AL, AR ]   = FLA_Part_1x2( A,  0, 'FLA_LEFT' );
  [ ATT, ABT ] = FLA_Part_2x1( AT, 0, 'FLA_TOP'  );

  while size(AL,2) < size(A,2)

      b = min( size(AR,2), nb_alg );

      % Pull a b-column panel from A  (m × b)
      [ A0, A1, A2 ] = FLA_Repart_1x2_to_1x3( AL, AR, b, 'FLA_RIGHT' );

      % Pull the corresponding b-row panel from AT  (b × m)
      [ AT0, AT1, AT2 ] = FLA_Repart_2x1_to_3x1( ATT, ABT, b, 'FLA_BOTTOM' );

      % Update C with that panel product
      C = Syrk_ln_unb_var5( A1, AT1, C );

      % Slide partitions
      [ AL, AR ]   = FLA_Cont_with_1x3_to_1x2( A0,  A1,  A2, 'FLA_LEFT'  );
      [ ATT, ABT ] = FLA_Cont_with_3x1_to_2x1( AT0, AT1, AT2, 'FLA_TOP'   );
  end

  C_out = C;
end

File: projectfiles/flamec/syrk_ln_blk_var5.c
/* Copyright 2025 The University of Texas at Austin  
 
   For licensing information see
                  http://www.cs.utexas.edu/users/flame/license.html 

   Programmed by: Umer Khan
                  umerkhan5110@gmail.com
*/


#include "FLAME.h"

FLA_Error syrk_ln_blk_var5( FLA_Obj A, FLA_Obj C, dim_t nb_alg )
{
  FLA_Obj AL, AR, A0, A1, A2;
  dim_t b;

  FLA_Part_1x2( A, &AL, &AR, 0, FLA_LEFT );

  while ( FLA_Obj_width( AL ) < FLA_Obj_width( A ) ) {

    b = min( nb_alg, FLA_Obj_width( AR ) );

    FLA_Repart_1x2_to_1x3( AL, /**/ AR, &A0, /**/ &A1, &A2,
                           b, FLA_RIGHT );

    /*------------------------------------------------------------*/

    /* C := C + A1 * A1ᵀ */
    FLA_Syrk( FLA_LOWER_TRIANGULAR,
              FLA_NO_TRANSPOSE,
              FLA_ONE, A1,
              FLA_ONE, C );

    /*------------------------------------------------------------*/

    FLA_Cont_with_1x3_to_1x2( &AL, /**/ &AR, A0, A1, /**/ A2,
                              FLA_LEFT );
  }

  return FLA_SUCCESS;
}
